SimPEG/potential_fields/gravity/_numba_functions.py

```diff
@@ -0,0 +1,177 @@
+"""
+Numba functions for gravity simulation using Choclo.
+"""
+import numpy as np
+
+try:
+    import choclo
+except ImportError:
+    # Define dummy jit decorator
+    def jit(*args, **kwargs):
+        return lambda f: f
+
+    choclo = None
+else:
+    from numba import jit, prange
+
+
+def _forward_gravity(
+    receivers,
+    nodes,
+    densities,
+    fields,
+    cell_nodes,
+    kernel_func,
+    constant_factor,
+):
+    """
+    Forward model the gravity field of active cells on receivers
+
+    This function should be used with a `numba.jit` decorator, for example:
+
+    ..code::
+
+        from numba import jit
+
+        jit_forward_gravity = jit(nopython=True, parallel=True)(_forward_gravity)
+        jit_forward_gravity(
+            receivers, nodes, densities, fields, cell_nodes, kernel_func, const_factor
+        )
+
+    Parameters
+    ----------
+    receivers : (n_receivers, 3) array
+        Array with the locations of the receivers
+    nodes : (n_active_nodes, 3) array
+        Array with the location of the mesh nodes.
+    densities : (n_active_cells)
+        Array with densities of each active cell in the mesh.
+    fields : (n_receivers) array
+        Array full of zeros where the gravity fields on each receiver will be
+        stored. This could be a preallocated array or a slice of it.
+    cell_nodes : (n_active_cells, 8) array
+        Array of integers, where each row contains the indices of the nodes for
+        each active cell in the mesh.
+    kernel_func : callable
+        Kernel function that will be evaluated on each node of the mesh. Choose
+        one of the kernel functions in ``choclo.prism``.
+    constant_factor : float
+        Constant factor that will be used to multiply each element of the
+        ``fields`` array.
+
+    Notes
+    -----
+    The constant factor is applied here to each element of fields because
+    it's more efficient than doing it afterwards: it would require to
+    index the elements that corresponds to each component.
+    """
+    n_receivers = receivers.shape[0]
+    n_nodes = nodes.shape[0]
+    n_cells = cell_nodes.shape[0]
+    # Evaluate kernel function on each node, for each receiver location
+    for i in prange(n_receivers):
+        # Allocate vector for kernels evaluated on mesh nodes
+        kernels = np.empty(n_nodes)
+        for j in range(n_nodes):
```

Comment:
I am guessing this is where the parallel C process happens? Otherwise it would be really slow to loop - this could technically be done with linear algebra right?

Reply:
All these functions are decorated with `@numba.jit`, so Numba is compiling these functions using LLVM and then using these compiled versions when they are called.

If by linear algebra you mean vectorized functions, it's possible, although those operations involve some level of abstractions that usually result in less efficient code (allocation of large arrays, unnecessary allocation of memory for intermediate results, etc). That's why I went with plain for loops.

Does it make sense?

Reply:
yea it does. So the loops are turned into C loops, which is basically what numpy does (I think?) anyway.
Thanks